nav2_amcl/src/amcl_node.cpp

```diff
@@ -20,440 +20,219 @@
+
+  // Get rid of the inputs first (services and message filter input), so we
+  // don't continue to process incoming messages
+  global_loc_srv_.reset();
+  nomotion_update_srv_.reset();
+  initial_pose_sub_.reset();
+  laser_scan_connection_.disconnect();
+  laser_scan_filter_.reset();
+  laser_scan_sub_.reset();
+
+  // Map
+  map_free(map_);
+  map_ = nullptr;
+  free_space_indices.resize(0);
 
-  handleMapMessage(*msg);
-  first_map_received_ = true;
-}
+  // Transforms
+  tf_broadcaster_.reset();
+  tf_listener_.reset();
+  tf_buffer_.reset();
 
-void
-AmclNode::handleMapMessage(const nav_msgs::msg::OccupancyGrid & msg)
-{
-  std::lock_guard<std::recursive_mutex> cfl(configuration_mutex_);
+  // PubSub
+  pose_pub_.reset();
+  particlecloud_pub_.reset();
 
-  RCLCPP_INFO(get_logger(), "Received a %d X %d map @ %.3f m/pix\n",
-    msg.info.width,
-    msg.info.height,
-    msg.info.resolution);
-  if (msg.header.frame_id != global_frame_id_) {
-    RCLCPP_WARN(
-      get_logger(), "Frame_id of map received:'%s' doesn't match global_frame_id:'%s'. This could"
-      " cause issues with reading published topics",
-      msg.header.frame_id.c_str(),
-      global_frame_id_.c_str());
-  }
+  // Odometry
+  motion_model_.reset();
 
-  freeMapDependentMemory();
-  // Clear queued laser objects because they hold pointers to the existing
-  // map, #5202.
+  // Particle Filter
+  pf_free(pf_);
+  pf_ = nullptr;
+
+  // Laser Scan
   lasers_.clear();
   lasers_update_.clear();
   frame_to_laser_.clear();
 
-  map_ = convertMap(msg);
-
-#if NEW_UNIFORM_SAMPLING
-  // Index of free space
-  free_space_indices.resize(0);
-  for (int i = 0; i < map_->size_x; i++) {
-    for (int j = 0; j < map_->size_y; j++) {
-      if (map_->cells[MAP_INDEX(map_, i, j)].occ_state == -1) {
-        free_space_indices.push_back(std::make_pair(i, j));
-      }
-    }
-  }
-#endif
-  // Create the particle filter
-  pf_ = pf_alloc(min_particles_, max_particles_,
-      alpha_slow_, alpha_fast_,
-      (pf_init_model_fn_t)AmclNode::uniformPoseGenerator,
-      reinterpret_cast<void *>(map_));  // (void *)map_);
-  pf_->pop_err = pf_err_;
-  pf_->pop_z = pf_z_;
-
-  // Initialize the filter
-  updatePoseFromServer();
-  pf_vector_t pf_init_pose_mean = pf_vector_zero();
-  pf_init_pose_mean.v[0] = init_pose_[0];
-  pf_init_pose_mean.v[1] = init_pose_[1];
-  pf_init_pose_mean.v[2] = init_pose_[2];
-  pf_matrix_t pf_init_pose_cov = pf_matrix_zero();
-  pf_init_pose_cov.m[0][0] = init_cov_[0];
-  pf_init_pose_cov.m[1][1] = init_cov_[1];
-  pf_init_pose_cov.m[2][2] = init_cov_[2];
-  pf_init(pf_, pf_init_pose_mean, pf_init_pose_cov);
-  pf_init_ = false;
-
-  // Instantiate the sensor objects
-  delete motionModel_;
-  createMotionModel();
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
 
-  // Laser
-  delete laser_;
-  createLaserObject();
+nav2_lifecycle::CallbackReturn
+AmclNode::on_error(const rclcpp_lifecycle::State &)
+{
```

Comment:
@crdelsey No default behavior. There is a pure virtual method that has to be overridden by the lifecycle node. Going forward, we can probably notify the lifecycle manager if we encounter an unrecoverable error and the lifecycle manager could respond by perhaps restarting the node. This is work to be done. For now, we can exit, but we should add this to all of the nodes so that they behave the same way. How about a separate PR for this?

Reply:
> For now, we can exit, but we should add this to all of the nodes so that they behave the same way. How about a separate PR for this?

That would be fine.

Reply:
Ok, I've added an issue for this: https://github.com/ros-planning/navigation2/issues/712